dali/core/mm/default_resources.cc

```diff
@@ -127,24 +128,110 @@ CallAtExit<Callable> AtExit(Callable &&c) {
+    if (UseDeferredDealloc()) {
+      using resource_type = mm::async_pool_resource<mm::memory_kind::device, cuda_vm_resource,
+                                                    std::mutex, void>;
+      return std::make_shared<resource_type>();
+    } else {
+      using resource_type = mm::async_pool_resource<mm::memory_kind::device, cuda_vm_resource_base,
+                                                    std::mutex, void>;
+      return std::make_shared<resource_type>();
+    }
+  }
+#endif
+  {
+    static auto upstream = std::make_shared<mm::cuda_malloc_memory_resource>();
+    if (UseDeferredDealloc()) {
+      using resource_type = mm::async_pool_resource<mm::memory_kind::device>;
+      auto rsrc = std::make_shared<resource_type>(upstream.get());
+      return make_shared_composite_resource(std::move(rsrc), upstream);
+    } else {
+      using resource_type = mm::async_pool_resource<mm::memory_kind::device,
+              pool_resource_base<memory_kind::device, any_context, coalescing_free_tree, spinlock>>;
+      auto rsrc = std::make_shared<resource_type>(upstream.get());
+      return make_shared_composite_resource(std::move(rsrc), upstream);
+    }
+  }
 #endif
 }
 
+class simple_pinned_resource : public mm::async_memory_resource<memory_kind::pinned> {
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    return upstream_.allocate(bytes, alignment);
+  }
+  void *do_allocate_async(size_t bytes, size_t alignment, stream_view stream) override {
+    return upstream_.allocate(bytes, alignment);
+  }
+  void do_deallocate(void *mem, size_t bytes, size_t alignment) override {
+    return upstream_.deallocate(mem, bytes, alignment);
+  }
+  void do_deallocate_async(void *mem, size_t bytes, size_t alignment, stream_view stream) override {
+    return upstream_.deallocate(mem, bytes, alignment);
+  }
+  rmm::mr::pinned_memory_resource upstream_;
+};
+
 inline std::shared_ptr<pinned_async_resource> CreateDefaultPinnedResource() {
-  using resource_type = mm::async_pool_resource<mm::memory_kind::pinned>;
+  if (!UsePinnedMemoryPool()) {
+    static auto upstream = std::make_shared<simple_pinned_resource>();
+    return upstream;
+  }
   static auto upstream = std::make_shared<rmm::mr::pinned_memory_resource>();
-  auto rsrc = std::make_shared<resource_type>(upstream.get());
-  return make_shared_composite_resource(std::move(rsrc), upstream);
+  if (UseDeferredDealloc()) {
+    using resource_type = mm::async_pool_resource<mm::memory_kind::pinned>;
+    auto rsrc = std::make_shared<resource_type>(upstream.get());
+    return make_shared_composite_resource(std::move(rsrc), upstream);
+  } else {
+    using resource_type = mm::async_pool_resource<mm::memory_kind::pinned,
+        pool_resource_base<memory_kind::pinned, any_context, coalescing_free_tree, spinlock>>;
+    auto rsrc = std::make_shared<resource_type>(upstream.get());
+    return make_shared_composite_resource(std::move(rsrc), upstream);
+  }
```

Comment:
Is there any naming scheme that explains what happens between:
`rmm::mr::cuda_memory_resource`, `mm::cuda_malloc_memory_resource`,
(and to lesser degree `mm::async_pool_resource`)?


And why I don't see `mm::async_memory_resource`? Is this because they are `rmm::mr::stream_ordered_memory_resource` and we don't use stream order yet?

Reply:
I use the name `async` simply because it's shorter than `stream_ordered`